From 6457960507fc052a4aedca4091d91b02f054331d Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Fri, 17 Jun 2016 14:05:47 +0100
Subject: [PATCH 337/348] Mythmusic: Provide network control of playlist

Allows querying the available playlist names and playing a named
playlist.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythplugins/mythmusic/mythmusic/musicplayer.cpp |   42 +++++++++++++++-
 mythtv/programs/mythfrontend/networkcontrol.cpp |   58 ++++++++++++++++++++++-
 2 files changed, 97 insertions(+), 3 deletions(-)

diff --git a/mythplugins/mythmusic/mythmusic/musicplayer.cpp b/mythplugins/mythmusic/mythmusic/musicplayer.cpp
index 0a7537b..be1025f 100644
--- a/mythplugins/mythmusic/mythmusic/musicplayer.cpp
+++ b/mythplugins/mythmusic/mythmusic/musicplayer.cpp
@@ -709,6 +709,31 @@ void MusicPlayer::customEvent(QEvent *event)
                              QString("MusicPlayer: got invalid MUSIC_COMMAND "
                                      "PLAY_TRACK - %1").arg(me->Message()));
                 }
+                else if (list[2] == "PLAY_LIST")
+                {
+                    if (list.size() < 4)
+                    {
+                        LOG(VB_GENERAL, LOG_ERR,
+                             QString("MusicPlayer: got invalid MUSIC_COMMAND "
+                                     "PLAY_LIST - %1").arg(me->Message()));
+                    }
+                    else if (gMusicData && gMusicData->all_playlists)
+                    {
+                        int start = me->Message().indexOf("'");
+                        int end = me->Message().lastIndexOf("'");
+                        if (start != -1 && end != -1 && start != end)
+                        {
+                            QString name = me->Message().mid(start + 1, end - start - 1);
+                            Playlist *plist = gMusicData->all_playlists->getPlaylist(name);
+                            if (plist)
+                            {
+                                gMusicData->all_playlists->copyToActive(plist->getID());
+                                qApp->processEvents();
+                                setCurrentTrackPos(0);
+                            }
+                        }
+                    }
+                }
                 else if (list[2] == "GET_METADATA")
                 {
                     QString mdataStr;
@@ -737,6 +762,20 @@ void MusicPlayer::customEvent(QEvent *event)
                     MythEvent me(message);
                     gCoreContext->dispatch(me);
                 }
+                else if (list[2] == "GET_PLAYLISTS")
+                {
+                    QString mdataStr;
+
+                    if (gMusicData && gMusicData->all_playlists)
+                        mdataStr = gMusicData->all_playlists->getPlaylistNames().join(", ");
+                    else
+                        mdataStr = "<Uninitialised>";
+
+                    QString message = QString("MUSIC_CONTROL ANSWER %1 %2")
+                            .arg(gCoreContext->GetHostName()).arg(mdataStr);
+                    MythEvent me(message);
+                    gCoreContext->dispatch(me);
+                }
             }
             else
                 LOG(VB_GENERAL, LOG_ERR,
@@ -1514,7 +1553,8 @@ void MusicPlayer::activePlaylistChanged(int trackID, bool deleted)
     if (!getCurrentMetadata())
     {
         m_currentTrack = -1;
-        stop(true);
+        if (isPlaying())
+            stop(true);
     }
 }
 
diff --git a/mythtv/programs/mythfrontend/networkcontrol.cpp b/mythtv/programs/mythfrontend/networkcontrol.cpp
index 4dfeac3..97c4d35 100644
--- a/mythtv/programs/mythfrontend/networkcontrol.cpp
+++ b/mythtv/programs/mythfrontend/networkcontrol.cpp
@@ -777,6 +777,37 @@ QString NetworkControl::processPlay(NetworkCommand *nc, int clientID)
                 message = QString("MUSIC_COMMAND %1 PLAY_FILE '%2'")
                                 .arg(hostname)
                                 .arg(nc->getFrom(3));
+            else if (is_abbrev("list", nc->getArg(2)))
+            {
+                gotAnswer = false;
+
+                MythEvent me(QString("MUSIC_COMMAND %1 GET_PLAYLISTS")
+                                .arg(gCoreContext->GetHostName()) );
+                gCoreContext->dispatch(me);
+
+                QTime timer;
+                timer.start();
+                while (timer.elapsed() < FE_SHORT_TO && !gotAnswer)
+                {
+                    qApp->processEvents();
+                    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+                }
+
+                if (gotAnswer && answer == "<Uninitialised>")
+                {
+                    GetMythMainWindow()->JumpTo(jumpMap["playlistview"]);
+
+                    QTime timer;
+                    timer.start();
+                    while ((timer.elapsed() < FE_LONG_TO) &&
+                           (GetMythUI()->GetCurrentLocation().toLower() != "playlistview"))
+                        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+                }
+
+                message = QString("MUSIC_COMMAND %1 PLAY_LIST '%2'")
+                                .arg(hostname)
+                                .arg(nc->getFrom(3));
+            }
             else
                 return QString("ERROR: Invalid 'play music' command");
         }
@@ -1055,6 +1086,27 @@ QString NetworkControl::processQuery(NetworkCommand *nc)
             return QString("ERROR: See 'help %1' for usage information "
                            "(parameters mismatch)").arg(nc->getArg(0));
     }
+    else if (is_abbrev("playlists", nc->getArg(1)))
+    {
+        gotAnswer = false;
+
+        MythEvent me(QString("MUSIC_COMMAND %1 GET_PLAYLISTS")
+                        .arg(gCoreContext->GetHostName()) );
+        gCoreContext->dispatch(me);
+
+        QTime timer;
+        timer.start();
+        while (timer.elapsed() < FE_SHORT_TO && !gotAnswer)
+        {
+            qApp->processEvents();
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+
+        if (gotAnswer)
+            return answer;
+
+        return "unknown";
+    }
     else if (is_abbrev("events", nc->getArg(1)))
     {
         result = nc->getClient()->WantsEvents() ? "On" : "Off";
@@ -1465,7 +1517,8 @@ QString NetworkControl::processHelp(NetworkCommand *nc)
             "play music getstatus   - Get music player status playing/paused/stopped (MythMusic)\r\n"
             "play music file NAME   - Play specified file (MythMusic)\r\n"
             "play music track N     - Switch to specified track (MythMusic)\r\n"
-            "play music url URL     - Play specified URL (MythMusic)\r\n";
+            "play music url URL     - Play specified URL (MythMusic)\r\n"
+            "play music list NAME   - Play specified playlist (MythMusic)\r\n";
     }
     else if (is_abbrev("query", command))
     {
@@ -1485,7 +1538,8 @@ QString NetworkControl::processHelp(NetworkCommand *nc)
             "query loglevel        - Get current VERBOSE loglevel\r\n"
             "query version         - Query Frontend version details\r\n"
             "query channels        - Query available channels\r\n"
-            "query channels START LIMIT - Query available channels from START and limit results to LIMIT lines\r\n";
+            "query channels START LIMIT - Query available channels from START and limit results to LIMIT lines\r\n"
+            "query playlists       - Query playlists (MythMusic)\r\n";
     }
     else if (is_abbrev("set", command))
     {
-- 
1.7.9.5

